pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java

```diff
@@ -90,64 +90,72 @@ enum MetadataValueType {
+    NUM_ENTRIES_SCANNED_POST_FILTER(4, "numEntriesScannedPostFilter", MetadataValueType.LONG),
+    NUM_SEGMENTS_QUERIED(5, "numSegmentsQueried", MetadataValueType.INT),
+    NUM_SEGMENTS_PROCESSED(6, "numSegmentsProcessed", MetadataValueType.INT),
+    NUM_SEGMENTS_MATCHED(7, "numSegmentsMatched", MetadataValueType.INT),
+    NUM_CONSUMING_SEGMENTS_QUERIED(8, "numConsumingSegmentsQueried", MetadataValueType.INT),
+    NUM_CONSUMING_SEGMENTS_PROCESSED(26, "numConsumingSegmentsProcessed", MetadataValueType.INT),
+    NUM_CONSUMING_SEGMENTS_MATCHED(27, "numConsumingSegmentsMatched", MetadataValueType.INT),
+    MIN_CONSUMING_FRESHNESS_TIME_MS(9, "minConsumingFreshnessTimeMs", MetadataValueType.LONG),
+    TOTAL_DOCS(10, "totalDocs", MetadataValueType.LONG),
+    NUM_GROUPS_LIMIT_REACHED(11, "numGroupsLimitReached", MetadataValueType.STRING),
+    TIME_USED_MS(12, "timeUsedMs", MetadataValueType.LONG),
+    TRACE_INFO(13, "traceInfo", MetadataValueType.STRING),
+    REQUEST_ID(14, "requestId", MetadataValueType.LONG),
+    NUM_RESIZES(15, "numResizes", MetadataValueType.INT),
+    RESIZE_TIME_MS(16, "resizeTimeMs", MetadataValueType.LONG),
+    THREAD_CPU_TIME_NS(17, "threadCpuTimeNs", MetadataValueType.LONG),
+    SYSTEM_ACTIVITIES_CPU_TIME_NS(18, "systemActivitiesCpuTimeNs", MetadataValueType.LONG),
+    RESPONSE_SER_CPU_TIME_NS(19, "responseSerializationCpuTimeNs", MetadataValueType.LONG),
+    NUM_SEGMENTS_PRUNED_BY_SERVER(20, "numSegmentsPrunedByServer", MetadataValueType.INT),
+    NUM_SEGMENTS_PRUNED_INVALID(21, "numSegmentsPrunedByInvalid", MetadataValueType.INT),
+    NUM_SEGMENTS_PRUNED_BY_LIMIT(22, "numSegmentsPrunedByLimit", MetadataValueType.INT),
+    NUM_SEGMENTS_PRUNED_BY_VALUE(23, "numSegmentsPrunedByValue", MetadataValueType.INT),
+    EXPLAIN_PLAN_NUM_EMPTY_FILTER_SEGMENTS(24, "explainPlanNumEmptyFilterSegments", MetadataValueType.INT),
+    EXPLAIN_PLAN_NUM_MATCH_ALL_FILTER_SEGMENTS(25, "explainPlanNumMatchAllFilterSegments", MetadataValueType.INT);
+
+    private static final MetadataKey[] ID_TO_ENUM_KEY_MAP;
     private static final Map<String, MetadataKey> NAME_TO_ENUM_KEY_MAP = new HashMap<>();
+
+    private final int _id;
     private final String _name;
     private final MetadataValueType _valueType;
 
-    MetadataKey(String name, MetadataValueType valueType) {
+    MetadataKey(int id, String name, MetadataValueType valueType) {
+      _id = id;
       _name = name;
       _valueType = valueType;
     }
 
-    // getByOrdinal returns an enum key for a given ordinal or null if the key does not exist.
+    /**
+     * Returns the MetadataKey for the given id, or {@code null} if the id does not exist.
+     */
     @Nullable
-    public static MetadataKey getByOrdinal(int ordinal) {
-      return VALUES[Math.min(ordinal, VALUES.length - 1)];
+    public static MetadataKey getById(int id) {
+      return id < ID_TO_ENUM_KEY_MAP.length ? ID_TO_ENUM_KEY_MAP[id] : null;
```

Comment:
This appears to be an unnecessary change as there is a null element at the end of the map to deal with unmapped ordinals.

Reply:
Functionality wise, it doesn't make a difference (I updated it to handle negative input as well). Readability wise, IMO it is more readable without extra overhead. It does take me some time to understand that we append a null element to the end of the array.

Reply:
Ok, it’s better now it supports negative values